packages/drivers/odsp-driver/src/fetcher.ts

```diff
@@ -0,0 +1,92 @@
+import { IsoBuffer } from "@fluidframework/common-utils";
+import { OdspErrorType } from "./odspError";
+import { fetchAndParseAsBufferHelper, fetchAndParseAsJSONHelper, IOdspResponse } from "./odspUtils";
+
+export class Fetcher {
+    constructor(
+        private fluidEpoch: string | undefined,
+        private readonly logger?: ITelemetryLogger) {
+    }
+
+    public async fetchAndParseAsJSON<T>(
+        url: string,
+        fetchOptions: {[index: string]: any},
+        addInBody: boolean = false,
+    ): Promise<IOdspResponse<T>> {
+        // Add epoch either in header or in body.
+        this.addEpochInRequest(fetchOptions, addInBody);
+        let response: IOdspResponse<T>;
+        try {
+            response = await fetchAndParseAsJSONHelper<T>(url, fetchOptions);
+            this.extractEpochFromResponse(response.headers);
+        } catch (error) {
+            this.checkForEpochError(error);
+            throw error;
+        }
+        return response;
+    }
+
+    public async fetchAndParseAsBuffer<T>(
+        url: string,
+        fetchOptions: {[index: string]: any},
+        addInBody: boolean = false,
+    ): Promise<IOdspResponse<IsoBuffer>> {
+        // Add epoch either in header or in body.
+        this.addEpochInRequest(fetchOptions, addInBody);
+        let response: IOdspResponse<IsoBuffer>;
+        try {
+            response = await fetchAndParseAsBufferHelper(url, fetchOptions);
+            this.extractEpochFromResponse(response.headers);
+        } catch (error) {
+            this.checkForEpochError(error);
+            throw error;
+        }
+        return response;
+    }
+
+    private addEpochInRequest(fetchOptions: {[index: string]: any}, addInBody: boolean) {
+        if (this.fluidEpoch !== undefined) {
+            if (addInBody) {
+                // We use multi part form request for post body where we want to use this.
+                // So extract the form boundary to mark the end of form.
+                let body: string = fetchOptions.body;
+                const formBoundary = body.split("\r\n")[0].substring(2);
+                body += `\r\nepoch=${this.fluidEpoch}\r\n`;
+                body += `\r\n--${formBoundary}--`;
+                fetchOptions.body = body;
+                return;
+            }
+            // Else add in headers.
+            fetchOptions.headers = {
+                ...fetchOptions.headers,
+                "x-fluid-epoch": this.fluidEpoch,
+            };
+        }
+    }
+
+    private extractEpochFromResponse(headers: Map<string, string>) {
+        const epochFromResponse = headers.get("x-fluid-epoch");
+        // If epoch is undefined, then don't compare it because initially for createNew or TreesLatest
+        // initializes this value. Sometimes response does not contain epoch as it is still in
+        // implementation phase at server side. In that case also, don't compare it with our epoch value.
+        assert(this.fluidEpoch === undefined
```

Comment:
When we learn about mismatch in epochs, we have to take proactive step and clean cache entries for this file.
If we do not do it, then we will continue to load from cache using old epoch, and on any operation on a file will learn about new epoch and fail - and get into same exact state on refresh.

Reply:
Added logic to remove all entries for such document from cache on mismatch.